File: menu-ocr-android/app/src/main/java/com/menuocr/viewmodel/AuthViewModel.kt
package com.menuocr.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.menuocr.data.AuthRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import io.github.jan.supabase.gotrue.user.UserInfo
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class AuthViewModel @Inject constructor(
    private val authRepository: AuthRepository
) : ViewModel() {

    private val _authState = MutableStateFlow<AuthState>(AuthState.Idle)
    val authState: StateFlow<AuthState> = _authState

    init {
        checkCurrentUser()
    }

    private fun checkCurrentUser() {
        val user = authRepository.currentUser
        _authState.value = if (user != null) {
            AuthState.Authenticated(user)
        } else {
            AuthState.Unauthenticated
        }
    }

    fun signInWithEmail(email: String, password: String) {
        viewModelScope.launch {
            _authState.value = AuthState.Loading
            val result = authRepository.signInWithEmail(email, password)
            result.onSuccess {
                _authState.value = AuthState.Authenticated(it)
            }.onFailure {
                _authState.value = AuthState.Error(it.message ?: "Sign in failed")
            }
        }
    }

    fun signUpWithEmail(email: String, password: String) {
        viewModelScope.launch {
            _authState.value = AuthState.Loading
            val result = authRepository.signUpWithEmail(email, password)
            result.onSuccess {
                _authState.value = AuthState.Authenticated(it)
            }.onFailure {
                _authState.value = AuthState.Error(it.message ?: "Sign up failed")
            }
        }
    }

    fun signInWithGoogle() {
        viewModelScope.launch {
            _authState.value = AuthState.Loading
            val result = authRepository.signInWithGoogle()
            result.onSuccess {
                _authState.value = AuthState.Authenticated(it)
            }.onFailure {
                _authState.value = AuthState.Error(it.message ?: "Google sign in failed")
            }
        }
    }

    fun signOut() {
        viewModelScope.launch {
            _authState.value = AuthState.Loading
            val result = authRepository.signOut()
            result.onSuccess {
                _authState.value = AuthState.Unauthenticated
            }.onFailure {
                _authState.value = AuthState.Error(it.message ?: "Sign out failed")
            }
        }
    }

    fun resetPassword(email: String) {
        viewModelScope.launch {
            _authState.value = AuthState.Loading
            val result = authRepository.resetPassword(email)
            result.onSuccess {
                _authState.value = AuthState.PasswordResetSent
            }.onFailure {
                _authState.value = AuthState.Error(it.message ?: "Password reset failed")
            }
        }
    }
}

sealed class AuthState {
    object Idle : AuthState()
    object Loading : AuthState()
    object Unauthenticated : AuthState()
    object PasswordResetSent : AuthState()
    data class Authenticated(val user: UserInfo) : AuthState()
    data class Error(val message: String) : AuthState()
}

File: menu-ocr-ios/MenuOCR/MenuOCR/ViewModels/AuthViewModel.swift
//
//  AuthViewModel.swift
//  MenuOCR
//
//  Created by MenuOCR on 2025-01-07.
//

import Foundation
import Combine

class AuthViewModel: ObservableObject {
    @Published var authState: AuthState = .idle

    private let supabaseService = SupabaseService.shared

    init() {
        checkCurrentUser()
    }

    private func checkCurrentUser() {
        if let user = supabaseService.getCurrentUser() {
            authState = .authenticated(user)
        } else {
            authState = .unauthenticated
        }
    }

    func signInWithEmail(email: String, password: String) {
        authState = .loading

        Task {
            do {
                let user = try await supabaseService.signInWithEmail(email: email, password: password)
                await MainActor.run {
                    authState = .authenticated(user)
                }
            } catch {
                await MainActor.run {
                    authState = .error(error.localizedDescription)
                }
            }
        }
    }

    func signUpWithEmail(email: String, password: String) {
        authState = .loading

        Task {
            do {
                let user = try await supabaseService.signUpWithEmail(email: email, password: password)
                await MainActor.run {
                    authState = .authenticated(user)
                }
            } catch {
                await MainActor.run {
                    authState = .error(error.localizedDescription)
                }
            }
        }
    }

    func signInWithGoogle() {
        authState = .loading

        Task {
            do {
                let user = try await supabaseService.signInWithGoogle()
                await MainActor.run {
                    authState = .authenticated(user)
                }
            } catch {
                await MainActor.run {
                    authState = .error(error.localizedDescription)
                }
            }
        }
    }

    func signOut() {
        authState = .loading

        Task {
            do {
                try await supabaseService.signOut()
                await MainActor.run {
                    authState = .unauthenticated
                }
            } catch {
                await MainActor.run {
                    authState = .error(error.localizedDescription)
                }
            }
        }
    }

    func resetPassword(email: String) {
        authState = .loading

        Task {
            // Note: Password reset would need to be implemented in SupabaseService
            await MainActor.run {
                authState = .passwordResetSent
            }
        }
    }
}